### DISCUSSION TITLE
New Features

### DISCUSSION THREAD

**Discussion by Haiwei1027** [Captain-Robertson/Stormworks---Bungeling-Raiders-Hostile-A.I.-#12] [Ideas]
Let's use this space to brainstorm new features.

**Comment by Haiwei1027** [comment]
Here's some ideas I've picked out from the steam workshop comment section:
## Land vehicles
We can probably easily do land enemies that don't move much
or use existing land pathfinding solution
## Multiplayer 
specifically other players missing map marker
## Air vehicles 
We can probably do helicopters easily
good fighter AI might be challenging
## Toggle if AI can have missiles
we can probably do this by adding missile tag to boats and filter them out
## Ability to set AI size preference 
e.g. only small boat pirates
we can probably do this by adding separate limit for each AI size

**Reply by Captain-Robertson** [reply]
I like the idea of the size preferences and the AI missile toggle, ships already have a size tag included and adding a missile tag would be easy enough to do.

Land vehicles are definitely an option, they'll require a little more work to implement though and I'll need to obtain/build a small selection of them, although there's some we can already take from the Weapons DLC.

Aircraft are going to be more challenging though, as we'll need to give them code similar to what's in the Weapons DLC to make them perform attack runs on the player once they get within a certain range. Right now the AI in this addon just move between waypoints.
If we do decide to add these then maybe having a separate toggle button on game start and slider for them might be a good idea, as they would be much more challenging for newer players, especially given how quickly they can move around.

The other players missing the map marker in Multiplayer was because I'd had the server.addMapObject set to 0 (own player/server owner) instead of -1 (all Peers) in the initial version I released on the workshop, as I wasn't sure how much lag it would cause. I've since changed it to -1 and so far there's been no more reports of missing markers or lag when using the addon, so I'm hoping it's all good now?

**Reply by Haiwei1027** [reply]
I think we should take little steps to avoid more bug fixes in the future. So, for land and air I think we should prototype the bare minimum and see how to take it forward from there.

**Reply by Captain-Robertson** [reply]
Agreed.

Taking things slowly and one small step at a time has always been my mantra and what you've managed to fix and help us get working so far has made the addon incredibly well polished compared to what it was before, so it's definitely the way to go. 
Even Casserole - a friend of mine who normally moans about everything on the workshop - sent me a message saying how smooth this addon runs and professional it looks, along with how he's going to use it in all his playthroughs from now on, which is high praise indeed coming from him. ;)

**Reply by Haiwei1027** [reply]
Ooh nice! I also try to code with failsafe's in mind as I've experience plenty of issues with addons and crashes, so I try to write code that I think will last the longest even when unexpected things happen.

**Comment by Captain-Robertson** [comment]
I forgot to mention, Zizo messaged me earlier when I offered to add him as a collaborator on GitHub so he could pop in and take a look at our code but he came back with:

** "I have a github account but I've never used it and I'm not familiar at all with how it works. I saw the code on github is already modified, and the fix I wanted to suggest already implemented in some way.

I would still suggest not to use onVehicleDespawn to determine reward and all, I'd turn that into a custom function that you can call from just above line 408. Determining vehicle data after the vehicle has despawned hasn't always worked consistently in different game updates." **

I wasn't aware there was any issues with using onVehicleDespawn as a trigger but given how insanely experienced Zizo is when it comes to working with Stormworks it might be worth considering if we run into any problems with it going forward?

**Reply by Haiwei1027** [reply]
I'll keep that in mind, currently my solution to the double reward relies on determining vehicle data in onVehicleDespawn(). But for future we can switch to his suggestion if we bump into issues in that aspect.

**Comment by Haiwei1027** [comment]
With all of the initial bugs fixed and enhancements implemented along with some minor enhancements. I think we should plan for the next stage.
Here are my opinions on the requested features from the comment section:

## AI Fleet/Convoy
With friendly fire issue resolved, this should be simple enough as we can assign a group index to vehicles and have vehicles with the same group index share destinations. Although size will have to be taken into consideration as multiple large ships in a group will lag most players.
## Land AI
Static land enemy will be very simple, but that may be unsatisfactory to players that enjoy this addon for its difficulty. We can probably take existing land pathing solution from workshop as I don't have hours or weeks to place land pathing nodes.
## Aircraft AI
Helicopters should be pretty simple as they don't require complicated attack sequence, but planes would need dedicated attack state unless they are gunships with turrets.
I could implement a new state for the AIs where if a target enters its range it switch to attack state and actively chase or circle the victim. It won't be hard at all since code like that already exists in weapon DLC AI addon. It would also make ship combat even more engaging.
## Submarine
This probably requires the least amount of code, yet most amount of design and building, so I don't really have an opinion on this.
# Here are some minor features I suggest:
## Protection fee
The player can pay the raiders protection fee for a period of peace.
## Chat message
Have the raiders send chat message from time to time asserting their power or messages to threaten the player

**Comment by Captain-Robertson** [comment]
Some good ideas there!

If the submarine we're experimenting with works out well enough and doesn't have any glaring issues then we can definitely consider maybe adding an attack helicopter or VTOL into the mix, as you've mentioned they'll have turrets and just circle and strafe so they should hopefully be easier to code than fast, fixed wing aircraft that need complicated attack runs 'etc'.

**Comment by Captain-Robertson** [comment]
I've just added the 'enhancement' of Vehicle Groups to the issues section based on Haiwei's excellent suggestion of AI Fleet/Convoys, especially now that the latest release is exceptionally stable.

This is definitely low priority, more so as we near the festive season and our offline commitments (especially for those of us that are still in full time education or have evening work) are on the increase, but it would be a nice surprise for our existing followers and subscribers out there - along with the possibility of land vehicles being added at some point too.

Regardless, that's some excellent work that's been done already and this addon has well and truly exceeded my initial expectations, so a huge thanks to everyone involved - including all the play testers out there that are reading this but haven't been added as official contributors!

**Comment by Captain-Robertson** [comment]
I've just added the first experimental iteration of fixed-wing (jet) aircraft into the mod.

Currently they are based on the old Sea Harrier and AV8-B Harrier II jump jets, so they use thrust vectoring to help prevent them from nose diving into the sea during the steep attack runs (gun_run is AI state 2 for aircraft which is different to the AI state 3 used by helos), and I've given them a microcontroller that automatically switches on a rudimentary autopilot when it doesn't detect the AI is manipulating the controls for more than 3 seconds - which will usually stop them from plummeting into the ground after they succeed in destroying their current target and the AI "zones out" and enters into their "waiting period" (was there a reason why it was made as long as 1 minute? i.e. line 576 where it reads 60 * 60 * 1 ticks?).

At the moment I've only added two variants of the aeroplanes, a bomber that uses twin heavy autocannon and two large rocket pods, plus a strike fighter variant that uses twin light autocannons (I may switch this to a single rotary autocannon if it's possible), two small rocket pods and six radar guided missiles, just so that the player still has the option to fight against one type of aircraft without missiles being enabled in the menu.

Currently they only engage in the gun_run type of behaviour, firstly because auto aiming turrets struggle with high velocities on the parent vehicle and secondly because of the inherent jerkiness of the hard-coded AI plane pilots making any fixed weapons miss all over the place, let alone any turrets that need a few seconds to stabilise. 
During tests adding gyros and PIDs just seemed to make the AI pilots understeer and fail the gun_run entirely, so this was eventually abandoned.

They're not perfect and the AI Pilots may still occasionally crash into the sea and/or terrain and will sometimes even manage to shoot themselves with their own rockets or gunfire when manoeuvring sharply, but this has always been the case even with the default AI aircraft from the Weapons DLC (in fact they are significantly worse) and has been drastically reduced by the use of MC's and overall aircraft design.

If all goes well and I'm relatively happy with how they perform then hopefully I'll add them into the next official release of this addon.

Best,
